### ISSUE TITLE
Firefox - two m= audio lines when renegotiating

### ISSUE THREAD

**Issue by rscreene** [onsip/SIP.js#456]
In the code below I intially start a voice only call to FreeSWITCH, I then call reinvite() to add video to the call.  On Firefox 59.0 the re-INVITE has a single m= video line and two m= audio lines (one with a port of zero).

On Chrome there is only a single m= audio line which is what I would expect.

This is on OSX 10.10.1

```
<script src="https://cdn.rawgit.com/onsip/SIP.js/0.8.2/dist/sip-0.8.2.min.js"></script>
<script>
var wsServers = [{"ws_uri":"wss://vhost.work/websocket","sip_uri":"<sip:vhost.work;transport=ws;lr>","weight":0,"status":0,"scheme":"WSS"}];

var sipUserAgent = new SIP.UA({ wsServers: wsServers, register: false, traceSip: true, log: { builtinEnabled: true } });

var sipSession;

function start() {
  var uri = "sip:link@vhost.work";

  var sipOptions = {
    sessionDescriptionHandlerOptions: {
      constraints: {
        audio: true,
        video: false
      }
    }
  };

  console.log("sipOptions=", sipOptions);
  sipSession = sipUserAgent.invite(uri, sipOptions);
  console.log("sipSession=", sipSession);

  sipSession.sessionDescriptionHandler.on("userMedia", function (stream) {
    console.log("GOT gUM STREAM", stream);
    document.getElementById("local").srcObject = stream;
  });


  sipSession.sessionDescriptionHandler.on("addStream", function (streamEvent) {
    console.log("GOT STREAM", streamEvent);

    document.getElementById("remote").srcObject = streamEvent.stream;

    sipSession.sessionDescriptionHandler.peerConnection.getRemoteStreams()[0].onaddtrack = function (trackEvent) {
      console.log("adding track", trackEvent);
    };
  });
}

function upgrade() {
  var sipOptions = {
    sessionDescriptionHandlerOptions: {
      constraints: {
        audio: true,
        video: true
      }
    }
  };

  console.log("sipOptions=", sipOptions);
  console.log("sipSession=", sipSession);
  sipSession.reinvite(sipOptions);
}


</script>

<button onclick="start();">Start!</button>
<button onclick="upgrade();">Upgrade!</button>

<video id="remote" autoplay></video>
<video id="local" autoplay muted="muted"></video>
```
[reneg.log](https://github.com/onsip/SIP.js/files/1414270/reneg.log)

**Comment by egreenmachine** [comment]
Welcome back! I'm glad to see you are testing out our recent changes.

I think I see the issue on [Line 152](https://github.com/onsip/SIP.js/blob/master/src/WebRTC/SessionDescriptionHandler.js#L152). When you reinvite with the video upgrade, we ask the user for permission again. Once that is complete we then take the tracks and apply them to the Peer Connection. Since you already had audio and now have audio and video this time around audio is added again. Chrome appears to detect and handle the duplicates, but Firefox does not. I would expect similar results if you had video only then upgraded to audio/video. Where the audio in that case would only have 1 m-line and the video to have 2 m-lines. Also a downgrade may not remove the tracks.

It's tough to know what to do here. I'm afraid that if I remove all the tracks from the Peer Connection if the renegotiation fails we won't have anything to fall back on and the call would fail. I may be able to detect the difference then just add or remove what has changed. This still suffers from the same problem of not having a fallback, but it seems like an acceptable middle ground. I will look into a patch, but it may take a few days as this requires a bit of testing across multiple browsers.

**Comment by jlaine** [comment]
Fyi this situation also seems to occur when calling hold() on an audio-only call with Firefox. PR #466 has a nice side-effect which is that we don't call getUserMedia and hence don't re-add the tracks, but the root cause seems to be the same as for this issue.

**Comment by egreenmachine** [comment]
This is resolved since #571 was merged
